### ISSUE TITLE
Question: Why C# doesn't allow passing parameter as final?

### ISSUE THREAD

**Issue by AhmedRagheb** [dotnet/csharplang#1126]
In Java, you can mark variable as final

```
public void GetById(final int id)
{
    // Get by id and don't change id 
}
```
Why this is not allowed in C#.
There are many questions like this in Stackoverflow but I prefer to ask here to get the right reason behind that, not just theoretical answers.

**Comment by mikedn** [comment]
> That's coming in C# 7.2, using in.

Nope. `in int` is really `ref readonly int`, that's not the same as `readonly int`.

**Comment by DavidArno** [comment]
@mikedn,

How is it different in practice?

**Comment by mikedn** [comment]
> How is it different in practice?

Hmm, that should be obvious. `ref readonly int` passes an `int` by reference and prevents you from modifying it. In 99.99% of the cases doing that is completely pointless. And slower too. And makes the method signature incompatible with one that passed the int by value.

`readonly int` simply prevents the code inside the method to modify the parameter. It doesn't change the signature, it doesn't affect performance etc.

I would hope that people don't start using `in int` as a surrogate for `readonly int`, that would be pretty obnoxious.

**Comment by Joe4evr** [comment]
Also, take into account that if you're passing a struct type (like `int`) *not* by `ref`, then changes to that parameter aren't observable at the caller because it's a value copy.

**Comment by DavidArno** [comment]
> I would hope that people don't start using `in int` as a surrogate for `readonly int`, that would be pretty obnoxious.

Ah, that's unfortunate as that is exactly what I plan on doing. I don't really care what the compiler does under the hood. `in` let's me mark a parameter as read-only, so I fully intend using it that way. 😇

**Comment by mikedn** [comment]
> Ah, that's unfortunate as that is exactly what I plan on doing. I don't really care what the compiler does under the hood

I hope I'll never have to read/update code written by you :grin:

**Comment by DavidArno** [comment]
@mikedn,

Just to clarify, your position is:
1. `in` is possibly slower than a dedicated `readonly` modifier,
2. There is no `readonly` modifier planned for 7.2,
3. `in` still shouldn't be used as "read-only" because it might be slower and/or because it's somehow "obnoxious" to do so.

Questions:
1. Do you have any evidence that it is slower? Have you measured it?
2. Surely worrying about that potential slowness is simple micro-optimisation?
3. In what way (other than possibly making the code slower, is it "obnoxious" to use a feature that (almost) exists, as opposed to waiting for a feature that might never happen?

**Comment by VisualMelon** [comment]
@DavidArno 
I think mikedn's position also includes the _considerably more important_ stance that because `in` is actually a reference it can change, and that means the caller can break their code in exiting ways they don't realise; I'll add that it's even worse than that, because there is nothing at the call site telling you that you are passing an lvalue by reference: _this is an obnoxious concept and I still can't believe they are going with it!!!_.

I can't think of many reasons not to add a `readonly` modifier (which has no change of semantics at the call site, unlike `in`) to improve method implementations without fear of the `in` keyword and its numerous subtleties. People will of course tell me that hardly anyone will use it because `in` is shorter to type (you'll be unsurprised to hear that I don't think much of arguments that appeal to the length of tokens). Another reason not to add readonly, of course, is that then `ref readonly` would look like it should exist, and that could create confusion (I'd rather it did exist, so that I don't have to go anywhere near `in` ever).

Regarding Question 3, waiting would be a problem because `in` will be in the wild and ruining lives (this might be hyperbole), but people will learn to use it and then won't use `readonly` if it were to be added (more over, it never would because people will - as you have - suggest using `in` instead). Personally, I would be very glad to see `readonly` added (with or without a call-site signature, no change in semantics means the call-site doesn't care, so it could be completely backward compatible); I think the `in` keyword is a disaster.

I'm sorry this turned into another rant...

**Comment by Joe4evr** [comment]
> Another reason not to add readonly, of course, is that then `ref readonly` would look like it should exist

But [`ref readonly` *does* exist](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRKgEwPYDsA2AnuhGADSYgDUAPgLABQAAgMwAETAjAGwcBMbAMJtGAb0ZtJHdl15MALGwCyACiRg2GHAUJsAlrhj6AlCIZS2481IC+EqfcmsOPDooByKg/sMmzFqws7BhsgA=) - `in` is just an alias for both modifiers.

**Comment by VisualMelon** [comment]
@Joe4evr indeed, I meant only to refer to the syntax, to suggest that people might ask "why can't I type `ref readonly`" (which would be a reasonable question) to which the answer would be "it's called `in`". It's just another little surprise that I think would be worth avoiding, though it's rather a petty concern.

**Comment by VisualMelon** [comment]
Some points about performance (which I agree is pretty irrelevant; but, as I understand it `in` is being added primarily to support code that is required to be fast: it seems kind of reasonable to hold it to its own account):

An `in` parameter may be changed by a different thread: this may well have performance implications when writing threaded code, as memory barriers could prevent optimisations that would otherwise be made when assuming single threaded execution. I haven't spent much time looking at generated assembly recently, but I presume that a `ref` lookup also requires an additional indirection, so repeated access (again, in a multi-threaded context) _could_ be slower, though in practise this is what CPUs are good at so probably not.

Really, any external call could modify an `in` parameters target. I don't know how for certain `ref`s are presently treated in this instance, but presumably `in` will work the same way (either by producing undefined behaviour at the whim of the JIT, or aggressively looking up 'invalidated' `ref`s (I would assume this is the case; note that no 'real' memory barrier is warranted)).

I vaguely recall reading that when passing an rvalue to `in`, it makes a copy at the call-site and passes a reference (rather than JITing `2^N` versions of the method, or something unscaleable like that), so these `ref` concerns will always be present, and indeed there will be an added overhead at the call-site for rvalues (equal, I should think, to the cost of passing a ByVal param), though I can imagine that this will usually be optimised away).

Again, I only really care about the semantic concerns which throw up these same performance concerns, and I doubt that legitimate use of `in` will regularly coincide with a need for blazing performance _and_ intricate multi-threading (the call-site is mislead, none-the-less, which is always bad).

**Comment by mikedn** [comment]
> In what way, other than possibly making the code slower, is it "obnoxious" to use a feature that (almost) exists, as opposed to waiting for a feature that might never happen?

Well, I don't know, the simple fact that `in` is part of the method signature should be enough to rise eyebrows. Any gratuitous use of a language feature in ways it wasn't designed for is obnoxious. Even more so if it turns out that the language feature is not well understood.

**Comment by DavidArno** [comment]
@mikedn,

> Well, I don't know, the simple fact that `in` is part of the method signature should be enough to rise eyebrows.

`readonly` would be part of the method signature too.

**Comment by DavidArno** [comment]
@VisualMelon,
> I think mikedn's position also includes the *considerably more important* stance that because `in` is actually a reference it can change, and that means the caller can break their code in exciting ways they don't realise;

Doesn't the fact that `in` parameters can be changed less than existing parameters negate this concern? 
* a non-readonly struct will have a copy made before passing by reference,
* a read-only struct can't be changed anyway,
* reference types are already passed by reference, so can already be changed.
* by using `in`, you prevent re-assignment, ie less change than before.

**Comment by mikedn** [comment]
> which I agree is pretty irrelevant

Irrelevant to whom? In what context? It may be irrelevant if used occasionally/accidentally by people who do not understand the feature. It's not going to be irrelevant if it used on large scale. For example, if all method parameters in CoreFX (that can be changed to `in`) would be changed to `in` then you can be sure that the performance will sink.

> An in parameter may be changed by a different thread

You don't even need to involve threads in this. Method A has an `in` parameter. Method A calls method B that happens to modify the location reference by A's `in` parameter. Code in method A doesn't account for that case and shoots itself in the foot.

> as memory barriers could prevent optimisations that would otherwise be made when assuming single threaded execution

Again you do not need to involve threading into this. The simple fact that passing a local variable via an `in` parameter results in the address of the local variable being taken is enough to cause optimization issues, especially in JIT compilers.

**Comment by mikedn** [comment]
> readonly would be part of the method signature too.

Who says that? A design where `void foo(readonly int)` is different from `void foo(int)` is a broken design.

**Comment by DavidArno** [comment]
@mikedn,

Ah, I see what you mean now. So `void foo(in int)` and `void foo(in int)` would be different signatures from an overloading point of view. Overloading-wise, `void foo(readonly int)` and `void foo(int)` would be the same. The former simply adds a contract of behaviour to the signature.

I'm starting to agree with @VisualMelon: `in` is a mistake. Providing both `readonly` and `readonly ref` at the same time would have avoided confusion.

**Comment by mikedn** [comment]
> I'm starting to agree with @VisualMelon: in is a mistake. Providing both readonly and readonly ref at the same time would have avoided confusion.

IMO they should stick to the `ref readonly` syntax. While `in` makes sense as a complement to the existing `ref` and `out`, it seems that it is prone to misinterpretation. Also, I don't think that there will too many good uses for `ref readonly` so the shorter syntax is not warranted.

**Comment by benaadams** [comment]
@DavidArno

Given 
```csharp
private void DoThing(int current)
{
    for (; current < 500; current++)
    {
        Console.WriteLine(current);
    }
}
```
What benefit do you see either `in` or `readonly` having?

```csharp
private void DoThing(in int current)
{
    int localCurrent = current;
    for (; localCurrent < 500; localCurrent++)
    {
        Console.WriteLine(localCurrent);
    }
}
```

```csharp
private void DoThing(readonly int current)
{
    int localCurrent = current;
    for (; localCurrent < 500; localCurrent++)
    {
        Console.WriteLine(localCurrent);
    }
}
```

**Comment by VisualMelon** [comment]
@mikedn 
> Irrelevant to whom?

You are right, there are of course people with legitimate concerns. I believe David and I are of the same mind that meaningful semantics are of an order of magnitude of greater important than performance: I think the semantic argument for the proposal is sound, I don't see a reason to discuss performance. 

>You don't even need to involve threads in this

Indeed, this was expressed in my third paragraph. I have discussed the terrifying subtleties of `in` in great details elsewhere: I didn't want to go into another rant here! Again, I think the performance concerns are - I find myself wanting to say 'irrelevant' again, but it isn't the right word - insignificant compared to the semantic concerns with `in`. My post about performance was just to try to express that performance is _not_ simple as something of a rebuttal to David's questions: one would be hard pressed to find interesting code that they can predict the performance of without investigating in a case by case basis both the method and call-site: I do see this as an argument in favour of an independent `readonly`, but as I said above, I really wouldn't care about performance if they rethought `in`, but I would want `readonly` _non_-refs either way, because they are semantically different.

**Comment by DavidArno** [comment]
@benaadams,

I see it as bringing clarity to the intent of the code. I'd already write that as:
```cs
private void DoThing(int initialValue)
{
    for (var i = initialValue; i< 500; i++)
    {
        Console.WriteLine(i);
    }
}
```
as I do not feel that `current` is a meaningful parameter name, nor do I want to overload the meaning of it as being both the initial value and current value.

Being able to write:
```cs
private void DoThing(in int initialValue)
{
    for (var i = initialValue; i< 500; i++)
    {
        Console.WriteLine(i);
    }
}
```
then adds compiler checking against me accidentally modifying `initialValue`.

Whether it's `in` or `readonly` doesn't much worry me. But shortly, I'll be able to use `in` and may never be able to use `readonly`.

**Comment by benaadams** [comment]
> then adds compiler checking against me accidentally modifying `initialValue`.

So its not to do with the caller or the method signature as seen by the caller. 

You want local `readonly` values (including the parameters)?

i.e. `int` vs `readonly int` are irrelevant to the caller and add no information.

**Comment by DavidArno** [comment]
@benaadams,
> You want local `readonly` values (including the parameters)?

Yes. But I'm not getting them; I'm getting `in`. So I'll use that instead.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/188

**Comment by benaadams** [comment]
`in` is very different semantically and in behaviour than `readonly`

**Comment by VisualMelon** [comment]
@DavidArno 
>Doesn't the fact that in parameters can be changed less than existing parameters negate this concern?

I figure you've already worked out what I'm going to say, but I'll say it anyway for posterity! ([This is one of my other rants on the semantic concerns with `in`, which is very relevant](https://github.com/dotnet/csharplang/issues/945#issuecomment-340223880): I can't seem to find my other rants, I've really bad at searching for things)

Sadly no: while it _does_ prevents the implementer from doing anything stupid with the parameter _in the usual case_ (which is the purpose of this proposal) it doesn't protect the caller from abusing the power for passing a reference. `in` adds so much complexity that any certainty in the unchanging state of a `readonly` parameter has to be throw away. The sad reality is that they can change in exciting and unexpected/difficult to debug ways (not "whoops, I changed the param, I'd better delete that line from the method I'm stepping through!" but "Why is the param changing while I wait on some IO?!??! Where are my threads?!??!!"). Further, while reference types can always change their internal state, an `in object` can also change which object it is looking at, which is a whole different kettle of fish (it could change to a different implementing/derived type!) that is not expressed by the word `in`, and isn't expressed at all to the caller. The same goes for a struct, immutable or otherwise: what the `ref` points to can be overwritten. It is just unhelpful handing off the `readonly` parameter to a `ref` type parameter, because in trying to protect the implementer from themselves (in already easy to debug ways), you have opened up a myriad of opportunities for the caller to ruin your day (in really hard to debug ways).

@benaadams is right that `readonly` non-`ref` is semantically equivalent to the caller as non-`readonly` non-`ref`, which is really important. I _do_ want `readonly` params and `readonly` locals, for all the reasons that David has outlined (local expression of intent and compiler checking for the _implementer_, not the caller). This is much of why this idea makes so much sense: it is simple, unambiguous, and helpful. On the other hand, `in` is about as complicated as any feature could hope to be and fails to express _anything_ of this to the caller. (You'll have guessed by now that I really hate `in`)

**Comment by mikedn** [comment]
> Yes. But I'm not getting them; I'm getting in. So I'll use that instead.

If after all this lengthy discussion you still insist in using `in int` as a substitute for `readonly int` then the only thing I can do is to wish you good luck. You'll need it.

**Comment by benaadams** [comment]
> an in object can also change which object it is looking at, which is a whole different kettle of fish

passing objects `in` is bugged https://github.com/dotnet/roslyn/issues/23327

**Comment by DavidArno** [comment]
@benaadams, isn't that already covered by #1123? Not sure it's a bug from CyrusNajmabadi's comments.

**Comment by DavidArno** [comment]
@VisualMelon 
> Further, while reference types can always change their internal state, an in object can also change which object it is looking at, which is a whole different kettle of fish (it could change to a different implementing/derived type!)

Ah. Of course, I hadn't though about it that way around. My nicely behaved method, that undertakes to not re-assign variables, could have the value of that variable changed from under it.

Hmm, timed to rethink my use of `in`. 😬

**Comment by benaadams** [comment]
> isn't that already covered by #1123?

Not really; I don't understand what purpose it serves. But thank you for pointing me at the issue.

**Comment by VisualMelon** [comment]
@DavidArno I shouldn't wish to appeal to your human nature, but it means a lot to me that I've finally got my point across! (Even if it's too late!)

@benaadams it's because an `in` parameter simply prevents the method from changing the `ref` reference: it doesn't prevent the memory pointed at changing (i.e. an `object` reference). Similarly, it allows you to retrieve such an `object` reference, which you can then call any method on you like. Consider this code:

```
void MyMethod(in List<int> list)
{
    List<int> localList = list; // (1) same object reference, this compiles because we aren't modifying the ref
    localList.Add(1); // (2) obviously this is fine
    list.Add(1); // (3) ... but this looks weird
}

Even if calls to mutating methods were blocked (which is impossible at the moment, because 'mutating method' is undefined), e.g. on line (3), you could always copy away the object reference and call it (1) and (2). There is discussion going on about being able to define `readonly` methods and properties which are only accessible in a non-readonly context, but as long as you can pull a non-const reference out from an `in` parameter, this is the only behaviour that makes sense. `in` only stops you changing the reference, it doesn't stop you using it.

**Comment by HaloFour** [comment]
It's probably too late but I think I agree that `in` is an unfortunate keyword choice for that feature.  As an optimization intended for performance in specific circumstances and with a list of caveats I think it warrants the more verbose `readonly ref` qualifier.  The `in` keyword is simply way too ambiguous and doesn't remotely convey what it actually does, or that it impacts the actual type of the signature in any way.  I'd just assume that it was the counterpoint to `out`.  IIRC the goal with `in` was to make calling said methods less verbose, but then its use became entirely optional anyway.  I would've had no issue if calling a method with a `readonly ref` parameter could omit the `readonly ref` at the call site except when requiring disambiguation.

Anyway, this conversation was slightly hijacked by the whole `in` feature which is only barely related.  Being able to notate parameters and locals as `readonly` I think is a worthwhile feature on its own.  #188

**Comment by mikedn** [comment]
> It's probably too late but I think I agree that in is an unfortunate keyword choice for that feature.

AFAIK the feature is only in preview builds so it should not be too late. I for one I'm seriously considering opening an issue for that.

**Comment by VisualMelon** [comment]
@HaloFour 
>I would've had no issue if calling a method with a readonly ref parameter could omit the readonly ref at the call site except when requiring disambiguation.

Just to stay off track a moment longer... optional `in` (or `readonly ref`) it isn't good enough: it _has_ to be mandatory, else the caller hasn't acknowledge just how complicated the call is. The semantics a perfectly reasonable and well defined; the problem is that the keyword is terrible and the call-site doesn't have to acknowledge the _different_ semantics. There may be legitimates uses for changing `in` params, and for returning the references again, but what is not acceptable is the semantics of a 'normal', simple, and predictable ByVal can be confused at the call-site with the complicated (if well defined) `in` semantics. Even if programmers from now on _explicitly check_ for this trap (which I certainly don't want to do, and inexperienced programmers certainly won't), programmers in the past didn't, and their old code calling old APIs which are changed to use `in` will have silently changed semantics. If `in` was mandatory at the call-site I would shut up in an instant, even if I agree the keyword is a poor choice!

Back on track, I agree #188 is a valuable proposal either way.

**Comment by HaloFour** [comment]
@VisualMelon 

IIRC part of the reason they wanted it to be omitted was to better support scenarios where a modifier wouldn't be possible, such as the target of extension methods and as the operands of overloaded operators.  I don't think that would make it particularly necessary to allow omission of the modifiers in normal method calls, though.  I'm going to have to dig through the design notes on the feature to try to better understand the decisions behind this syntax.

**Comment by VisualMelon** [comment]
@HaloFour if you find out why they went with it, do tell me!! I have asked exactly that question already and have not been happy with any of the answers! Difficulties with operators etc. is, as you say, no excuse for ruining 'normal' method calls.

**Comment by tannergooding** [comment]
Anywhere you see `in` you can (roughly speaking) consider it equivalent to `const *` in C/C++:
* It is still a reference, so even if the callee can't modify the value directly
  * some other owner of the value can modify it
  * you can cast away the `readonly`/`const`-ness (using `System.Runtime.CompilerServices.Unsafe`)
* It can provide significantly better performance when passing value types
  * This can vary based on the underlying calling convention/ABI being used
  * On all major ABIs currently supported, this is (generally speaking) limited to structs larger than 16-bytes
    * Smaller structs don't (generally) see degradation from the required shadow-copying of the value to the stack
    * Vector and HVA values require special consideration, as they can be passed in register (default for System V, requires `__vectorcall` calling convention on Windows)

**Comment by HaloFour** [comment]
@tannergooding 

I don't think that anyone is questioning the use of the feature, just the syntax.  The `in` keyword does nothing to convey that there is a reference involved or that the signature is actually of a different type.  The feature itself requires quite advanced understanding of structs and the CLR to use correctly and will probably be fairly niche.

**Comment by MkazemAkhgary** [comment]
shouldn't `in` behavior be changed and instead of `ref readonly` becomes `readonly` instead? 

this wouldn't be a breaking change either. I mean whats the point of passing a value by `readonly ref`, what benefit or observable behavior does `ref` provide?

**Comment by tannergooding** [comment]
@MkazemAkhgary 

> It can provide significantly better performance when passing value types

**Comment by HaloFour** [comment]
@MkazemAkhgary 

> what benefit or observable behavior does `ref` provide?

Efficiency in passing a reference as opposed to a copy of a value type.  The ability to observe immediate changes to the value (relevant when dealing with arrays or fields).

What `readonly ref` does is allow for retaining the efficiency of `ref` parameters while also providing the compiler guarantee that the reference will not be modified.  As described in the motivation it is already fairly common in performance-heavy libraries like XNA for methods to accept `ref` parameters specifically to squeeze out additional performance, but being `ref` puts a lot of additional burden on the caller.

**Comment by benaadams** [comment]
@MkazemAkhgary a 4x4 Matrix of floats is 64 bytes and of doubles is 128 bytes; preforming a copy for each call is very inefficient an has a big impact when you only have 16ms per frame (11ms for VR)

**Comment by Pzixel** [comment]
I don't see why we even need this `in` keyword. It behave just like passing a variable itself. If it's because of performance reason (we don't make a copy, just passing what we have), why then people say that `in` performance is ever _worse_, than simple parameter one?

**Comment by mikedn** [comment]
> Why then not automatically use it for value types?

Because it doesn't actually provide significantly better performance... for **all** value types.

>  I don't see any difference between in int and int parameter.

Another piece of evidence that `in` is a bad keyword choice.

**Comment by benaadams** [comment]
> why then people say that in performance is ever worse, than simple parameter one?

For **small value types** its bad; unless you want the live link behviour (which is more-likely an unexpected side-effect)

`in int` passes a reference (8-bytes on x64) in CPU register; which then needs to be deferenced to get the value from stack memory every time it is used (as its a live link - not copy).

`int` passes a its actual value (4 bytes on x64) in a CPU register which can then be used directly (literally a copy in register, fastest memory on CPU).

For **large value types** its good

`in Matrix4x4` passes a reference (8-bytes on x64) in CPU register; which then needs to be deferenced to get the value from stack memory every time it is used (as its a live link - not copy).

`Matrix4x4` copies the struct to stack first; as its too large for register, then passes a reference (8-bytes on x64) in CPU register; which then needs to be deferenced to get the value from stack memory every time it is used (as its a live link - not copy, though would never change).


So for large value types it skips this bit `copies the struct to stack first`; for small value types it adds this bit ` deferenced to get the value from stack memory every time it is used`


**tl;dr** `in` is a speed up for large value-types and a slow down for small value types

**Comment by Pzixel** [comment]
@benaadams anyhow, why not then use it automatically when you can profit from it? It seems that the rule is simply `use ref if type length is smaller then X otherwise don't`.

**Comment by benaadams** [comment]
> why not then use it automatically when you can profit from it? 

Because they are different things
```csharp
public class Program
{
    private static int _value;

    private static void AddToValue()
    {
        _value += 5;
    }

    private static void PrintValueByParam(int value)
    {
        Console.WriteLine(nameof(PrintValueByParam));
        Console.WriteLine(value);
        AddToValue();
        Console.WriteLine(value);
    }
    private static void PrintValueByIn(in int value)
    {
        Console.WriteLine(nameof(PrintValueByIn));
        Console.WriteLine(value);
        AddToValue();
        Console.WriteLine(value);
    }


    public static void Main(string[] args)
    {
        _value = 0;
        PrintValueByParam(_value);
        _value = 0;
        PrintValueByIn(in _value);
    }
}
```
Outputs
```
PrintValueByParam
0
0
PrintValueByIn
0
5
```

**Comment by Pzixel** [comment]
@benaadams so this is a low-level unsafe feature for bit hacking. Am I right?

**Comment by HaloFour** [comment]
@Pzixel 

>  so this is a low-level unsafe feature for bit hacking. Am I right?

It's a "lower-level" safe feature for fast math.  Operations like this are common in libraries like XNA and they are currently burdensome to use in C# since to be performant the parameters are declared `ref` which makes them a lot harder to consume.

I would agree that it is somewhat niche compared to the C# market as a whole and does require a good degree of understanding on the part of the library authors in order to use correctly.

**Comment by benaadams** [comment]
Not really its to both imply intent and have compiler enforcement of it e.g. previously you'd be using `ref` here due to the performance difference:

Method            |        Mean |     Error |    StdDev |
------------------|-------------|-----------|-----------|
Matrix4x4ByValue  | 212.2870 ns | 1.2065 ns | 1.1285 ns |
Matrix4x4ByRef    | 159.5631 ns | 0.5604 ns | 0.5242 ns |

```
private static Vector4 DoMaths(ref Matrix4x4 matrix, Vector4 value)
```
However `ref` suggests the the called function will be changing the value; which is wrong on many levels; rather than just using the value.

So its so you can do:

```
private static Vector4 DoMaths(readonly ref Matrix4x4 matrix, Vector4 value)
```
Or the shorter `in`
```
private static Vector4 DoMaths(in Matrix4x4 matrix, Vector4 value)
```

Which implies the intent better and also enforces only read in the method.

Probably what many people are after is: "readonly for locals and parameters" https://github.com/dotnet/csharplang/issues/188 so you could do:

`void PrintValue(readonly int value)`

i.e. "readonly by value" rather than "readonly by reference"

**Comment by HaloFour** [comment]
@benaadams 

> Which implies the intent better and also enforces only read in the method.

How exactly does `in` imply the intent better?  I think that if anything this conversation demonstrates how ambiguous that term is.  I've found very little info in design meetings as to why `in` seems to be favored over the much more explicit `readonly ref`.  Sure, that's a lot more verbose, but if it only needs to be applied at the parameter declaration then that only impacts library authors which should represent the relative minority of code.

**Comment by benaadams** [comment]
@HaloFour 

> > Which implies the intent better and also enforces only read in the method.

> How exactly does `in` imply the intent better?

`readonly ref` vs `ref`

I assume `in` is for the triple of ref passing; all of which will change the signature and calling convention for the caller:
* `in` - read only by called function
* `ref` - read/write by called function
* `out` - write only by called function

`readonly` alone on the other hand is irrelevant to the caller and just semantic sugar

**Comment by benaadams** [comment]
The question I suppose is would `in` be as confusing if `readonly` existed? 👍 👎

**Comment by mikedn** [comment]
> The question I suppose is would in be as confusing if readonly existed? 

IMO it would be even more confusing.

Developer X:

> `in`?! `readonly`?! They look the same to me. I'll just use `in` cause it's shorter.

**Comment by HaloFour** [comment]
@benaadams 

It's a quartet.  We already have "no modifier" which already means "read only by called function" by virtue of being by value.  The new modifier means "read only reference by called function", which `in` doesn't remotely convey.  I understand that it may serve as a counterpoint to `out` which doesn't explicitly call out the fact that it is a `reference`, but that nature is implied by the fact that it couldn't actually write a value to the parameter affecting the caller in any other way.

> `readonly` alone on the other hand is irrelevant to the caller and only important inside the called function

I disagree.  `readonly` is critically important here as it is what allows the caller to significantly relax rules about what it may pass to that parameter, including `readonly` fields which would normally be verboten.  Again, I don't think that `in` conveys this.

That's really my only concern here.  Combined with the fact that this feature is relatively niche and requires a decent working knowledge of structs I don't think that the feature should be made super-easy to (mis)use.  I'm fine with the omission of the modifier from the call-site though.

**Comment by HaloFour** [comment]
@VisualMelon 

I'm not sure how much sense that makes given that you can pass expressions to the `readonly ref` parameter, and that it applies to scenarios where there's no where to put a modifier at all, such as the target of an extension method or an operand with an overloaded operator.

**Comment by benaadams** [comment]
@HaloFour 

> > `readonly` alone on the other hand is irrelevant to the caller and only important inside the called function

> I disagree. `readonly` is critically important here...

I mean to the caller, these are identical:
```csharp
void PrintValue(int value)
void PrintValue(readonly int value)
```
So the `readonly` is irrelevant to the caller.

**Comment by VisualMelon** [comment]
@HaloFour 
I've said many times (and I think I will have write all my complains about `in` into a single issue tonight) that I really couldn't care less about difficulties with extension methods and operators: there is _no excuse_ for leaving the most common type of method call ambiguous. I would simply not allow operators and extension methods (i.e. anywhere it is ambiguous for lack of syntax) to have `in` semantics: if you want performance, use a normal method call. `in` is _all about_ performance for the few that need it, not about confusing _everyone_ (including, even _especially_ the few) by allowing silent and painful to debug API changes.

You are right that it would look odd passing an expression to a `readonly ref`... so do something special with rvalues, implicitly passing a reference to an `lvalue` is a dreadful idea. Again, this is a feature targeted at those with serious performance concerns, if it's ugly in cases where it doesn't help (ignoring opportunity for dodgy compiler optimisations) then who cares?
